src/Mod/TechDraw/Gui/QGSPage.cpp

```diff
@@ -505,26 +698,95 @@ QGIView * QGSPage::addWeldSymbol(TechDraw::DrawWeldSymbol* weld)
             weldGroup->setFeature(weld);       //for QGIWS
             weldGroup->setViewFeature(weld);   //for QGIV
             weldGroup->updateView(true);
-        } else {
-            Base::Console().Error("QGVP::addWeldSymbol - no parent QGILL\n");
         }
-    } else {
-        Base::Console().Error("QGVP::addWeldSymbol - parent is not DV!\n");
     }
     return weldGroup;
 }
 
+void QGSPage::setDimensionGroups(void)
+{
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int dimItemType = QGraphicsItem::UserType + 106;
+
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == dimItemType) && (!(*itInspect)->group())) {
+            QGIView* parent = findParent((*itInspect));
+            if (parent) {
+                QGIViewDimension* dim = dynamic_cast<QGIViewDimension*>((*itInspect));
+                addDimToParent(dim,parent);
+            }
+        }
+    }
+}
+
+void QGSPage::setBalloonGroups(void)
+{
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int balloonItemType = QGraphicsItem::UserType + 140;
+
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == balloonItemType) && (!(*itInspect)->group())) {
+            QGIView* parent = findParent((*itInspect));
+            if (parent) {
+                QGIViewBalloon* balloon = dynamic_cast<QGIViewBalloon*>((*itInspect));
+                addBalloonToParent(balloon,parent);
+            }
+        }
+    }
+}
+
+void QGSPage::setLeaderGroups(void)
+{
+//    Base::Console().Message("QGSP::setLeaderGroups()\n");
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int leadItemType = QGraphicsItem::UserType + 232;
+
+    //make sure that qgileader belongs to correct parent.
+    //quite possibly redundant
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == leadItemType) && (!(*itInspect)->group())) {
+            QGIView* parent = findParent((*itInspect));
+            if (parent) {
+                QGILeaderLine* lead = dynamic_cast<QGILeaderLine*>((*itInspect));
+                addLeaderToParent(lead,parent);
+            }
+        }
+    }
+}
+
+void QGSPage::setRichAnnoGroups(void)
+{
+//    Base::Console().Message("QGSP::setRichAnnoGroups()\n");
+    const std::vector<QGIView *> &allItems = getViews();
+    std::vector<QGIView *>::const_iterator itInspect;
+    int annoItemType = QGraphicsItem::UserType + 233;
+
+    //make sure that qgirichanno belongs to correct parent.
+    for (itInspect = allItems.begin(); itInspect != allItems.end(); itInspect++) {
+        if (((*itInspect)->type() == annoItemType) && (!(*itInspect)->group())) {
```

Comment:
```suggestion
        if ((*itInspect)->type() == annoItemType && !(*itInspect)->group()) {
```
If possible, also avoid nesting. If possible, all these code parts should be gathered in a common function. Let's avoid shotgun surgery (https://en.wikipedia.org/wiki/Shotgun_surgery)